File: org-app/internal/db/tenant_db.go
package db

import (
	"database/sql"
	"fmt"
	"strings"

	_ "github.com/lib/pq"
)

func TenantDBName(tenantID string) string {
	return "org-app_" + tenantID
}

const orgSchema = `
CREATE TABLE IF NOT EXISTS org_units (
	id       SERIAL PRIMARY KEY,
	name     TEXT NOT NULL,
	parent_id INT REFERENCES org_units(id),
	created_at TIMESTAMPTZ DEFAULT NOW()
);
`

func CreateTenantDatabase(host, port, user, password, tenantID string) error {
	dbName := TenantDBName(tenantID)
	connStr := fmt.Sprintf("host=%s port=%s user=%s password=%s dbname=postgres sslmode=disable",
		host, port, user, password)
	db, err := sql.Open("postgres", connStr)
	if err != nil {
		return err
	}
	defer db.Close()
	if _, err := db.Exec("CREATE DATABASE \"" + dbName + "\""); err != nil {
		if !isDuplicateDB(err) {
			return err
		}
	}
	appConnStr := fmt.Sprintf("host=%s port=%s user=%s password=%s dbname=%s sslmode=disable",
		host, port, user, password, dbName)
	appDB, err := sql.Open("postgres", appConnStr)
	if err != nil {
		return err
	}
	defer appDB.Close()
	_, err = appDB.Exec(orgSchema)
	return err
}

func isDuplicateDB(err error) bool {
	if err == nil {
		return false
	}
	s := err.Error()
	return strings.Contains(s, "already exists") || strings.Contains(s, "duplicate")
}

// OpenTenantDB opens a connection to the logical DB for the given tenant.
func OpenTenantDB(host, port, user, password, tenantID string) (*sql.DB, error) {
	dbName := TenantDBName(tenantID)
	connStr := fmt.Sprintf("host=%s port=%s user=%s password=%s dbname=%s sslmode=disable",
		host, port, user, password, dbName)
	return sql.Open("postgres", connStr)
}

File: todo-app/lib/todo_app/tenant_db.ex
defmodule TodoApp.TenantDB do
  @moduledoc "Create tenant DB and open connections. DB name = todo-app_<tenant_id>"
  require Logger

  def db_name(tenant_id), do: "todo-app_#{tenant_id}"

  def create_tenant_database(tenant_id) do
    cfg = Application.get_env(:todo_app, :tenant_registry)
    host = to_charlist(cfg[:hostname])
    port = cfg[:port]
    user = to_charlist(cfg[:username])
    pass = to_charlist(cfg[:password] || "")
    db = db_name(tenant_id)

    opts = [hostname: host, port: port, username: user, password: pass, database: "postgres"]
    {:ok, conn} = Postgrex.start_link(opts)
    try do
      Postgrex.query!(conn, "CREATE DATABASE \"#{db}\"", [])
    rescue
      e in Postgrex.Error -> if e.postgres.code != :duplicate_database, do: reraise e, __STACKTRACE__
    end
    GenServer.stop(conn)

    schema_opts = [hostname: host, port: port, username: user, password: pass, database: db]
    {:ok, schema_conn} = Postgrex.start_link(schema_opts)
    Postgrex.query!(schema_conn, """
    CREATE TABLE IF NOT EXISTS todos (
      id SERIAL PRIMARY KEY,
      title TEXT NOT NULL,
      completed BOOLEAN DEFAULT FALSE,
      created_at TIMESTAMPTZ DEFAULT NOW()
    )
    """, [])
    GenServer.stop(schema_conn)

    TodoApp.TenantRegistry.register(tenant_id, to_string(cfg[:hostname]))
    :ok
  end

  def get_tenant_conn(tenant_id) do
    case TodoApp.TenantRegistry.get_host(tenant_id) do
      {:ok, host} -> open_conn(tenant_id, host)
      :error -> nil
    end
  end

  defp open_conn(tenant_id, host) do
    cfg = Application.get_env(:todo_app, :tenant_registry)
    opts = [
      hostname: to_charlist(host),
      port: cfg[:port],
      username: to_charlist(cfg[:username]),
      password: to_charlist(cfg[:password] || ""),
      database: db_name(tenant_id)
    ]
    case Postgrex.start_link(opts) do
      {:ok, pid} -> {:ok, pid}
      e -> e
    end
  end
end
